rfd/0095-okta-integration.md

```diff
@@ -0,0 +1,405 @@
+  for getting the list of applications assigned to a user. This will be diffed against the list
+  of all applications to determine which applications a user has access to. These results are
+  *not* paginated, so this is guaranteed to be one API call.
+* [Get user groups](https://developer.okta.com/docs/reference/api/users/#get-user-s-groups) for
+  getting the groups assigned to the user. These results are *not* paginated, so this is
+  guaranteed to be one API call.
+
+To configure the Okta provider, the following fields are expected in the application service
+configuration for the provider:
+
+```yaml
+use_uid_mapping: false # Set to true to use the UID mapping from the user's provider config.
+endpoint: https://my-org-endpoint.okta.com
+api-token: API-TOKEN-FROM-OKTA # https://developer.okta.com/docs/guides/create-an-api-token/main/
+```
+
+#### Mapping between Teleport users and Okta users
+
+If the provider is configured with `use_uid_mapping` to `false`, which is the default, the Okta
+provider will attempt to look for users in Okta using the user's Teleport username. If this is
+set to true, provider will look for a provider config `okta_user_id ` to determine which user ID
+to use for Okta user lookups.
+
+```yaml
+kind: user
+version: v2
+metadata:
+  name: alice
+spec:
+  roles: ['devs']
+  traits:
+    logins: ['admin']
+    kubernetes_groups: ['edit']
+  provider_configs:
+    okta:prod:
+      # Can be any user ID supported by Okta, documented in
+      # https://developer.okta.com/docs/reference/api/users/#get-user
+      okta_user_id: alice@mydomain.com
+```
+
+#### Launching Okta applications
+
+[Application objects from Okta](https://developer.okta.com/docs/reference/api/apps/#application-object) can return one or more `appLinks` which, when launched directly from
+Teleport, will forward to the expected application along with Okta authentication. These links
+will be used as the application URLs synced from Okta.
+
+**Note**: These links will be redirected to their actual external URLs rather than living behind
+Teleport's proxy.
+
+### Backend modifications
+
+A few modifications to existing services, databases, and models will be necessary to support this
+functinoality.
+
+#### Applications
+
+Applications will need some new information to support this:
+
+* A new flag to indicate that the application has been externally synchronized. This will allow
+  Teleport to distinguish between regular Teleport applications and external applications.
+* User specific application metadata for external applications. This will dictate to the UI what
+  kind of options to display and whether the URLs are accessible. These will be joined with the
+  application objects during query. To start, the only necessary fields for this object are a
+  user mapping, a corresponding external identity provider name, and a boolean flag as to
+  whether the user has access.
```

Comment:
This should also be controlled by RBAC. Our roles allow you to tell which other roles/resources you can request access to. E.g. check out https://goteleport.com/docs/access-controls/access-requests/resource-requests/?scope=enterprise.